lib/plugins/chat.js

```diff
@@ -243,4 +243,30 @@ function inject (bot, options) {
     })
   }
   bot.awaitMessage = awaitMessage
+
+  async function onMessage (onMessageTimeout, ...args) {
+    if (onMessageTimeout === undefined || args === undefined) {
+      throw new Error('Missing arguments!')
+    }
+    if (typeof Number(onMessageTimeout) !== 'number') {
+      throw new Error('Invalid onMessageTimeout value! Value must be an integer.')
+    }
+    onMessageTimeout = Number(onMessageTimeout)
+    return new Promise((resolve, reject) => {
+      const resolveMessages = args.flatMap(x => x)
+      function messageListener (msg) {
+        if (resolveMessages.some(x => x instanceof RegExp ? x.test(msg) : msg === x)) {
+          resolve(msg)
+          bot.off('messagestr', messageListener)
+          clearTimeout(timeoutOnMessage)
+        }
+      }
+      const timeoutOnMessage = setTimeout(async () => {
+        resolve(false)
+        bot.off('messagestr', messageListener)
+      }, onMessageTimeout)
+      bot.on('messagestr', messageListener)
+    })
+  }
+  bot.onMessage = onMessage
```

Comment:
Instead of making a new promise that resolver or rejects at some point you should use mineflayers [onceWithCleanup](https://github.com/PrismarineJS/mineflayer/blob/e5c7f7bad35a4778ade759297d4a63a8524c38a8/lib/promise_utils.js#L47) function. It also has a check condition to see if the promise should resolve.